uint8 right
uint8 left